src/middleware/logger.ts

```diff
@@ -39,3 +39,94 @@ export class DefaultLogger implements Logger {
     console.debug(...args);
   }
 }
+
+/**
+ * LogBuffer hold the args that will be passed on to the actual
+ * logger when attempting to flush.
+ */
+export class LogBuffer {
+  public readonly level: string;
+  public readonly args: LogArg[];
+
+  constructor(level: string, ...args: LogArg[]) {
+    this.level = level;
+    this.args = args;
+  }
+}
+
+/**
+ * ProxyLogger attempts to temporarily hold user's logs
+ * during the function run, and attempt to flush it through
+ * the provided logger when the function returns.
+ *
+ * The expected usage of this class is,
+ * 1. store the log attempt with its arguments and level
+ * 2. if function hits a step that has been memorized, clear the _buffer
+ * 3. on function return, flush all buffers
+ *
+ * And it should be invisible to the user as much as possible.
+ */
+export class ProxyLogger implements Logger {
+  private readonly _logger: Logger;
+  private _buffer: LogBuffer[] = [];
+
+  constructor(logger: Logger) {
+    this._logger = logger;
+  }
+
+  info(...args: LogArg[]) {
+    this._buffer.push(new LogBuffer("info", ...args));
+  }
+
+  warn(...args: LogArg[]) {
+    this._buffer.push(new LogBuffer("warn", ...args));
+  }
+
+  error(...args: LogArg[]) {
+    this._buffer.push(new LogBuffer("error", ...args));
+  }
+
+  debug(...args: LogArg[]) {
+    this._buffer.push(new LogBuffer("debug", ...args));
+  }
+
+  reset() {
+    this._buffer = [];
+  }
+
+  async flush() {
+    if (this.bufSize() === 0) return;
+
+    // eslint-disable-next-line @typescript-eslint/require-await
+    const deliveries = this._buffer.map(async (log) => {
+      const args = log.args;
+      const method = log["level"] as keyof Logger;
+      return this._logger[method](...args);
+    });
+    // NOTE: timestamp will likely not be linear as expected.
+    await Promise.all(deliveries);
```

Comment:
`Promise.all` will immediately reject on the first rejected promise in `deliveries`. Should we use `Promise.allSettled` instead?